compiler/optimizer/abstractinterpreter/AbsValue.hpp

```diff
@@ -0,0 +1,139 @@
+/*******************************************************************************
+ * Copyright (c) 2020, 2020 IBM Corp. and others
+ *
+ * This program and the accompanying materials are made available under
+ * the terms of the Eclipse Public License 2.0 which accompanies this
+ * distribution and is available at http://eclipse.org/legal/epl-2.0
+ * or the Apache License, Version 2.0 which accompanies this distribution
+ * and is available at https://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the
+ * Eclipse Public License, v. 2.0 are satisfied: GNU General Public License,
+ * version 2 with the GNU Classpath Exception [1] and GNU General Public
+ * License, version 2 with the OpenJDK Assembly Exception [2].
+ *
+ * [1] https://www.gnu.org/software/classpath/license.html
+ * [2] http://openjdk.java.net/legal/assembly-exception.html
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR Apache-2.0 OR GPL-2.0 WITH Classpath-exception-2.0 OR LicenseRef-GPL-2.0 WITH Assembly-exception
+ *******************************************************************************/
+
+#ifndef ABS_VALUE_INCL
+#define ABS_VALUE_INCL
+
+#include "il/OMRDataTypes.hpp"
+#include "compile/Compilation.hpp"
+#include "optimizer/VPConstraint.hpp"
+#include "optimizer/ValuePropagation.hpp"
+
+namespace TR { class AbsVPValue; }
+
+namespace TR {
+
+/**
+ * AbsValue is the abstract representation of a 'value'.
+ * It is the basic unit used to perform abstract interpretation.
+ * The Constraint needs to be implemented.
+ * One such example is VPConstraint.
+ */
+template<typename Constraint>
```

Comment:
I'm not sure the templating is needed here. If `_constraint` and `getConstraint()` are moved to a subclasses, then `AbsValue` doesn't need to know anything about how constraints are represented or stored. `setToTop()` and `isTop()` would have to become virtual, but that makes sense because a different constraint mechanism might have a different representation for Top than a null constraint. For `merge()`, the `other` parameter can be statically cast down to the correct type. Given that it's extremely unlikely for different kinds of constraints mechanism to get mixed together, the static down cast should be safe.

As a rough example:

```c++
class AbsValue
   {
   public:
   virtual TR::AbsValue* clone(TR::Region& region) = 0;
   virtual TR::AbsValue* merge(TR::AbsValue* other) = 0;
   virtual bool isTop() const = 0;
   virtual void setToTop() = 0;
   virtual void print(TR::Compilation* comp)=0;
   // ... other stuff mostly stays the same ...
   };

class AbsVPValue : public AbsValue
   {
   public:
   virtual TR::AbsValue* merge(TR::AbsValue* otherBase)
      {
      auto other = static_cast<TR::AbsVPValue*>(otherBase);
      // ... etc ....
      }

   // .. other stuff ...

   private:
   TR::VPConstraint* _constraint;
   };
```

Thoughts?

Reply:
@Leonardo2718  Yes, this is a good idea. I will modify it accordingly. 

But the trade-off of not using the template is there would be a lot of downcasting in abstract interpretation. For example, accessing the constraint.